### PR TITLE
[thorfinn] Round-2: Bilateral symmetry train-time augmentation (single delta)

### PR DESCRIPTION
**Pull request by morganmcg1**
# Round-2 Assignment: Bilateral symmetry training-time augmentation

## Hypothesis

DrivAerML cars are **exactly bilaterally symmetric** across the longitudinal
plane (y=0). Random horizontal mirroring during training is a free 2× expansion
of effective training data with zero distribution shift. Cars in the test set
are also symmetric, so this generalizes correctly.

This directly attacks the tau_y/tau_z disparity (current 21.80 / 23.07 vs
tau_x 15.73) — both axes contain antisymmetric components that the model has
not seen enough variation of within the 400-case training set.

This is **different from your previous bilateral TTA at eval time** (PR #37, now
closed): we augment training data, not test predictions. Train-time augmentation
exposes the model to mirrored geometries while learning, while eval-time TTA only
ensembles two predictions of an already-fixed model.

## Implementation

Add a `--bilateral-aug` flag (default off) in `train.py`. When enabled, in the
training-only `__getitem__` of the dataset class (in `data/loader.py` or wherever
the dataset class lives), with probability 0.5 per sample:

1. **Surface coords:** flip `y → -y` (i.e. `coords[:, 1] *= -1`)
2. **Volume coords:** flip `y → -y` similarly
3. **Targets — wall shear:** flip `tau_y → -tau_y` (the y-component of wall
   shear is a vector quantity that flips sign under reflection)
4. **Targets — wall shear x and z:** unchanged
5. **Targets — surface pressure, volume pressure:** unchanged (scalar fields)

Apply only to training split, NOT validation or test. Confirm by checking the
dataset transform / split flag.

If the dataset class is in `data/loader.py`, add the augmentation as a method
called from `__getitem__` only when `split == 'train'`.

### Sanity check before launching

After implementation, verify with a 1-epoch dry run that:
- Training loss is in the expected range (`~0.1-0.3` after first epoch)
- Val/test pipeline doesn't see flipped data
- Random seed is set so each rank gets a different flip pattern (or use a
  per-sample random rather than per-batch)

## Run command

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn \
  --wandb-name "thorfinn/round2-bilateral-aug-train-512d" \
  --wandb-group "tay-round2-bilateral-aug" \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --bilateral-aug \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

**Use compile (default).** PR #40 fixed the compile bug.

## What to report

After training completes, post a comment with all `test_primary/*` metrics:

| Metric | This run | PR #40 (SOTA) |
|---|---:|---:|
| `abupt` | ? | 17.25 |
| `surface_pressure` | ? | 10.92 |
| `wall_shear` | ? | 18.33 |
| `volume_pressure` | ? | 14.71 |
| `tau_x` | ? | 15.73 |
| `tau_y` | ? | 21.80 |
| `tau_z` | ? | 23.07 |

Specifically watch tau_y and tau_z — they should improve most (5-15% gain
expected). tau_x should be roughly flat. surface_pressure and volume_pressure
should be flat or slightly improved (more effective training data).

Also report:
- W&B rank-0 run ID
- Wall-clock time and number of epochs reached
- Best val_abupt
- Confirmation that augmentation was active (e.g. log a single sample's
  pre/post augmentation y-coord stats from rank-0 epoch 0).

## Critical: push code commits to remote!

Your previous PR #37 was closed because the per-axis-weights code was never
pushed despite the run completing. **For this PR, please push every code commit
to remote immediately after making it** so the diff is visible. Without the
diff, the PR cannot be merged.

```bash
git add data/loader.py train.py  # whatever you modified
git commit -m "feat: bilateral symmetry train-time augmentation"
git push origin thorfinn/round2-bilateral-aug-train
```

## Current baseline to beat

| Metric | tay best (PR #40) | yi best | AB-UPT |
|---|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **17.25** | 15.82 | — |
| `test_primary/surface_pressure_rel_l2_pct` | **10.92** | 9.99 | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **18.33** | 16.60 | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **14.71** | 14.21 | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **15.73** | 14.27 | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **21.80** | 19.49 | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **23.07** | 21.12 | 3.63 |

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #47 because ADVISOR (auto-posted, thorfinn pod hung post-train):

This run finished cleanly in W&B (`fn8pyav5`, runtime 5h 13m, best epoch via val) but the pod did not post results. Closing based on W&B-verified metrics — clear regression vs SOTA, no rebase or follow-up justified for this hypothesis as configured.

## Result

**test_abupt = 19.79** vs current SOTA 14.550 (**+36% regression**).

| Metric | PR #47 | tay SOTA (PR #46) | Δ |
|---|---:|---:|---:|
| `abupt` | **19.786** | 14.550 | **+36.0%** |
| `surface_pressure` | 13.258 | 8.628 | +53.7% |
| `wall_shear` | 21.107 | 14.882 | +41.8% |
| `volume_pressure` | 16.525 | 15.032 | +9.9% |
| `tau_x` | 18.524 | 12.901 | +43.6% |
| `tau_y` | 24.461 | 17.281 | +41.5% |
| `tau_z` | 26.163 | 18.907 | +38.4% |

Best val_abupt: 18.84. Wall-clock: 5h 13m (313 min).

## Diagnosis

Bilateral train-time augmentation introduced a regression on **every axis**. The mechanism likely failed because:

1. **Augmentation severity miscalibrated for DrivAerML's strong asymmetry signal.** The car geometries have weak left-right asymmetry (driver-side mirrors, exhaust placement, cabin layout), but augmenting symmetrically tells the model to expect identity behavior — which the loss penalizes when real test cases have those asymmetries.
2. **Tau_y/tau_z still failed** — the supposed target axes for symmetry exploitation. tau_y/tau_z are 41–38% worse than the new SOTA. So even on the metrics this hypothesis was designed to help, the result is much worse.

## Disposition: CLOSED

Bilateral symmetry as a pure augmentation does not generalize to DrivAerML's mild asymmetry distribution. If symmetry exploitation is worth revisiting, it should be at the **architecture level** (e.g., antisymmetric/symmetric heads or geometry-aware coordinate processing), not the augmentation level.

Will assign you a fresh hypothesis as soon as your pod recovers from the post-train hang (waiting on Issue #53 for restart)..
